layout: Unify size calculation of replaced elements

stshine opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14490)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/layout/construct.rs, components/layout/fragment.rs, components/layout/display_list_builder.rs, components/layout/block.rs, components/layout/flex.rs, components/layout/model.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify layout code, but no tests are modified. Please consider adding a test!

stshine commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 1c6b563 with merge 3c3bf82...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1304)

stshine commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 8d47987 with merge a81a1bb...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1306)

stshine commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 95a8e07 with merge da30026...

bors-servo commented:
:broken_heart: Test failed - [windows-dev](http://build.servo.org/builders/windows-dev/builds/2572)

stshine commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 95a8e07 with merge 864fafe...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1332)

asajeffrey reviewed (commented):
You probably need a reviewer who knows the layout system better than me!

emilio commented on components/layout/model.rs:
```diff
@@ -436,6 +436,21 @@ impl MaybeAuto {
     }
 }
 
+// Return used value for inline-size or block-size.
```
Can you make this a doc comment? Also, this doesn't return the used value (the used value is technically the _final_ pixel value).

Probably you can reword it somehow. This returns a middle-ground between the computed value and the used value.

stshine replied:
Sure, will do, this is from removed code.

emilio commented on components/layout/fragment.rs:
```diff
@@ -1203,6 +924,181 @@ impl Fragment {
+                    // First, have two rectangles that keep aspect ratio while may be clamped
+                    // by the contraints;
+                    let (first_isize, first_bsize) = {
+                        let isize = inline_constraint.clamp(intrinsic_inline_size);
```
Can you expand the name of `isize` so it isn't the same as a type? `inline_size` is fine, probably doing the same for `block_size`.

stshine replied:
There are already too many `inline_size`s, but I will found a better name.

emilio commented on components/layout/fragment.rs:
```diff
@@ -1203,6 +924,181 @@ impl Fragment {
+                    // Clamp the result and we are done :-)
+                    (inline_constraint.clamp(inline_size), block_constraint.clamp(block_size))
+                } else {
+                    (inline_constraint.clamp(intrinsic_inline_size),
```
I guess we can early return and deindent the rest? I think that'd help to keep it a bit more legible. Not a big deal though.

stshine replied:
Hmm... I am not sure about this, use two branch rather than an early return feels a clearer logic.

emilio replied:
I was mostly concerned about rightward drift, but it's ok.

emilio commented on components/layout/fragment.rs:
```diff
@@ -1203,6 +924,181 @@ impl Fragment {
+            SpecificFragmentInfo::Image(_)  |
+            SpecificFragmentInfo::Canvas(_) |
+            // TODO(stshine): According to the SVG spec, whether a SVG element has intrinsic
+            // aspect ratio is determined by the `preservAspectRatio` attribute. Since for now
```
nit: `preserveAspectRatio`.

emilio commented on components/layout/model.rs:
```diff
@@ -507,61 +522,59 @@ impl ToGfxMatrix for ComputedMatrix {
-pub struct MinMaxConstraint {
-    min: Au,
-    max: Option<Au>
+pub struct SizeConstraint {
```
(It'd be nice to keep the renaming in a different commit, but that doesn't matter now :P)

emilio commented on components/layout/model.rs:
```diff
@@ -507,61 +522,59 @@ impl ToGfxMatrix for ComputedMatrix {
     }
 
     /// Clamp the given size by the given `min` and `max` constraints.
+    /// Make sure the size is to be used for content box.
```
Can you elaborate on this comment? Technically this can be reused for other purposes.

stshine replied:
Will fix in a seperate commit.

emilio commented on components/layout/fragment.rs:
```diff
@@ -1549,66 +1444,39 @@ impl Fragment {
             }
-            SpecificFragmentInfo::Image(ref mut image_fragment_info) => {
-                let mut image_inline_size = match self.style.content_inline_size() {
+            SpecificFragmentInfo::Image(_) | SpecificFragmentInfo::Canvas(_) |
```
Can we move `is_replaced_content` to `Fragment` so we can reuse it here and everywhere else like:

```rust
_ if self.is_replaced_content() => {
   // ...
}
```

or alternatively:

```rust
if self.is_replaced_content() {
    // ....
    return ...;
}

match self.specific {
     SpecificFragmentInfo::Iframe(..) | Specifi... =>
         unreachable!("Replaced content should've been handled!"),
}
```

Since I guess we care about not forgetting to add fragment types to some parts of the layout code, and a compiler error is faster and better than a runtime error.

stshine replied:
Good idea, will do. There are so many repeated `is_replaced_or_*` code in `fragment.rs`.

emilio reviewed (commented):
This looks great! I love negative diffs :)

Some comments below.

emilio commented on components/layout/block.rs:
```diff
@@ -670,16 +670,16 @@ impl BlockFlow {
             SpecificFragmentInfo::Canvas(_) |
-            SpecificFragmentInfo::InlineBlock(_) => true,
+            SpecificFragmentInfo::Iframe(_) => true,
+            SpecificFragmentInfo::ScannedText(_) => panic!("A block that is also a text fragment!"),
```
@asajeffrey This is never supposed to happen, but I agree this panic is out of place.

stshine replied:
I thought I could complete this PR last night, but unfortunately ran into another bug: on replaced elements margins in block direction do not take effect. While this bug is really weird,  I will try to fix it this morning.

stshine commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 49dd0c5 with merge ed1b0e4...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1343)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14436) made this pull request unmergeable. Please resolve the merge conflicts.

stshine commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 70ae974 with merge 485cba9...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1338)

stshine commented:
Finally this is done, I'm so happy :D (Hope I will never need to deal with margin collapse and line metric at the same time again, even the simplest hacks.)

emilio commented:
Awesome, the new passes are exciting! Please update the tests expectations and tell me when you feel it's ready so I review it more in depth :)

stshine committed:
Rename the MinMaxConstraint to SizeConstraint and enhancement

Renamed to `SizeConstraint`, add an optional `border` parameter to
deal with `box-sizing: border-box`, and fix its bug when involving
with `calc`.

stshine committed:
Add new methods to Fragment for size calculation of replaced element

These new methods calculate both the used width and height of an
replaced element and the same time. The `has_intrinsic_ratio()` method
also exposes information about whether a fragment has intrinsic aspect
ratio.

stshine committed:
Switch to new methods for size calculation of replaced elements

Use `calculate_replaced_sizes()` in `compute_intrinsic_inline_sizes()`
and `assign_replaced_inline_size_if_necessary()`, and remove size
calculation in `assign_replaced_block_size_if_necessary()`, since
we've already done that.

stshine committed:
Remove unused replaced size calculation methods

Remove some fields and methods from SpecificFragmentInfo and change
some function signatures.

stshine committed:
Remove some unused methods

Simply use `Fragment::is_replaced()` in block, and remove
`content_inline_size()` method from Fragment.

stshine committed:
Various fixes

Fix the behavior of block-direction margins of replaced element both
as block level or inline level, and no longer query fragment info when
building display list for canvas.

stshine committed:
Update test expectations

stshine commented:
This pull request is ready for review.
And, hope you will like it :-)

emilio commented on components/layout/fragment.rs:
```diff
@@ -1202,6 +923,179 @@ impl Fragment {
+                            (first_isize, first_bsize),
+                        // When both rectangles grow (smaller than min sizes),
+                        // Choose the larger one;
+                        (Ordering::Greater, Ordering::Greater) if first_isize > second_isize =>
```
Probably this can be a bit clearer like:

```rust
(Ordering::Greater, Ordering::Greater) => {
    if first_isize > second_isize {
         // ...
    } else {
        // ...
    }
}
```

Instead of two different cases.

emilio commented on components/layout/fragment.rs:
```diff
@@ -1202,6 +923,179 @@ impl Fragment {
+        let block_constraint = self.size_constraint(containing_block_size, Direction::Block);
+
+        // https://drafts.csswg.org/css-images-3/#default-sizing
+        match (inline_size, block_size) {
```
Spec texts change, so in general more spec quoting on top of linking I think would be helpful.

This is mostly a personal preference, and I know this algorithm in particular isn't going to change, so I won't enforce it, feel free to add quotes when you feel it's appropriate.

stshine replied:
Added some spec texts.

emilio commented on components/layout/fragment.rs:
```diff
@@ -1548,66 +1441,41 @@ impl Fragment {
+                        self.border_padding.block_start += border.block_start;
+                        self.border_padding.block_end += border.block_end;
+                        let (isize, _) = self.calculate_replaced_sizes(None, None);
+                        isize
```
So I understand why you're doing this, but I think that it feels weird to have two places where we initialize border_padding. @pcwalton any thoughts?

stshine replied:
Yeah, this is done this way because 1. the `box_sizing_boundary()` gets the border info from the `border_padding` field, 2. the `surrounding_intrinsic_inline_size()` method does not return the block direction borders, 3. it is used during other layout passes other than intrinsic inline size bubbling so we can not simply store the border value in the `border_padding` field. If there is a good solution I'd love to take.

stshine replied:
Other the other hand, my next plan is to make an enhancement to our intrinsic inline size bubbling pass. Maybe I can attack this trouble together.

emilio replied:
Can you add a TODO there?

stshine replied:
Sure, will do

emilio reviewed (commented):
The rest looks great to me, this is a nice cleanup! :)

stshine commented:
Updated according to review comments.

emilio reviewed (commented):
Looks good to me, modulo that, though I think it's fine to land it and then rework how it works, over all if it's already on your radar.

r=me with that TODO/FIXME comment.

emilio commented:
@bors-servo delegate+

bors-servo commented:
:v: @stshine can now approve this pull request

stshine commented:
@bors-servo r+

@emilio Thanks :-)

bors-servo commented:
:pushpin: Commit 7bca303 has been approved by `stshine`

<!-- @bors-servo r=stshine 7bca3034b2d3d9aa2634d3ca5b7471fed254b770 -->

KiChjang commented:
@bors-servo r=emilio

stshine commented:
Oh no.
@bors-servo r-

stshine commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 7bca303 has been approved by `emilio`

<!-- @bors-servo r=emilio 7bca3034b2d3d9aa2634d3ca5b7471fed254b770 -->

bors-servo commented:
:hourglass: Testing commit 7bca303 with merge c7ecbfc...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4652)

KiChjang commented:
@bors-servo retry infra

bors-servo commented:
:hourglass: Testing commit 7bca303 with merge 9775d97...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1383)

stshine commented:
@bors-servo retry

infra

bors-servo commented:
:hourglass: Testing commit 7bca303 with merge 1993b6e...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/4587), [arm64](http://build.servo.org/builders/arm64/builds/4596), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4678), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1358), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1372), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4653), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4680), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1368), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1384), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2625)
